### PR TITLE
Fix memory leaks and other issues identified by the code sanitizer 

### PR DESCRIPTION
**Pull request by yantosca**
Using the `-DSANITIZE=y` CMake option (which invokes code sanitization) has identified several memory leaks and similar issues in HEMCO code. These issues can be caused by:

  - Arrays (or array fields of objects) that are allocated but not deallocated
  - Attempting to store a number in a variable that cannot accommodate it (i.e. trying to store 1e150 in a REAL*4 variable)
  - Attempting to store a value of the wrong type into a variable
  - etc.
 
This PR should be merged along with https://github.com/geoschem/GCClassic/pull/24 (which adds the proper compiler flags for the -DSANITIZE=y option) and https://github.com/geoschem/geos-chem/pull/1353 (which removes memory leaks etc. in the GEOS-Chem code).

For now, tagging @lizziel @msulprizio @Jourdan-He @jimmielin

### CODE REVIEW

**Comment by yantosca** [comment]
Also tagging @christophkeller

**Comment by yantosca** [comment]
Also see related issue #159

**Comment by lizziel** [comment]
This is great! Were you able to estimate the size of memory leak from the log before you added in the fixes?

**Comment by yantosca** [comment]
I am currently running integration tests with these updates merged up to HEMCO 3.5.1.  I can create a new PR.

**Comment by yantosca** [comment]
I have done a parallelization test with GEOS-Chem Classic using 8 cores vs. 5 cores.  The results show that the sanitize updates are zero-diff:

```console
﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿﻿Using configuration file rst.yml
... Printing totals and differences
... Only showing variables with |absolute difference| > 0.0
Variable                 Ref=5cores             Dev=8cores                Dev - Ref
```﻿

**Comment by yantosca** [comment]
Also updated the CHANGELOG.md

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,8 +5,14 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
-## Unreleased
-  - nothing yet
+## Unreleased [3.5.2]
+### Changed
+- Removed memory leaks that were identified by the code sanitizer
```
Comment (lizziel) [inline CHANGELOG.md:10]: I think the updates might be better under fixes: "Fixed memory leaks identified by code sanitizer"

File: CHANGELOG.md [hunk 2]
```diff
@@ -5,8 +5,14 @@ All notable changes to this project will be documented in this file.
...
+
```
Comment (lizziel) [inline CHANGELOG.md:11]: This block is part of 3.5.1. Maybe we should have a 3.5.2 to go with 14.0.2, so the target merge branch should be dev/3.5.2. That'll make these 3.5.1 updates go away in the file changes for this PR.

File: CHANGELOG.md [hunk 3]
```diff
@@ -5,8 +5,14 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
-## Unreleased
-  - nothing yet
+## Unreleased [3.5.2]
```
Comment (lizziel) [inline CHANGELOG.md:8]: The Changelog should also include under Added: "Added sanitizer option for detecting memory leaks in HEMCO standalone during build"

File: src/Extensions/hcox_dustdead_mod.F [hunk 4]
```diff
@@ -5638,74 +5638,149 @@ SUBROUTINE InstRemove ( Instance )
...
+            DEALLOCATE(Inst%SFC_TYP )
+         ENDIF
+         Inst%SFC_TYP => NULL()
+
+         IF ( ASSOCIATED( Inst%VAI_DST ) ) THEN
+            DEALLOCATE(Inst%VAI_DST )
+         ENDIF
+         Inst%VAI_DST => NULL()
+
+! These are not used
```
Comment (lizziel) [inline src/Extensions/hcox_dustdead_mod.F:5680]: Could these be deleted if not used?

**Review by lizziel — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (yantosca) [reply CHANGELOG.md:10]: Now updated in commit b404ae159df68506f4266529b0a6236ac416b1c9

File: CHANGELOG.md (hunk 2 above)
Comment (yantosca) [reply CHANGELOG.md:11]: I've pushed a dev/3.5.2 branch (currently at same commit as 3.5.1) and have specified that as the target branch for the merge.

File: CHANGELOG.md (hunk 3 above)
Comment (yantosca) [reply CHANGELOG.md:8]: Now added in commit: 471fa19e470a1b8a8c379305143bdbdaf5bdc945

File: src/Extensions/hcox_dustdead_mod.F (hunk 4 above)
Comment (yantosca) [reply src/Extensions/hcox_dustdead_mod.F]: Now removed in commit 758326e3d7886ff037502d871ac5c86b7ff6b1.  Changelog has been updated accordingly.

**Review by msulprizio — APPROVED** [review]

**Review by msulprizio — COMMENTED** [review]
These updates look good. I wonder if documentation should also be updated here for the sanitize option?

**Pull request merged**